samza-kafka/src/test/java/org/apache/samza/system/kafka/TestKafkaSystemConsumer.java

```diff
@@ -205,6 +215,125 @@ public void testFetchThresholdBytesDiabled() {
     consumer.stop();
   }
 
+  @Test
+  public void testStartpointSpecificOffsetVisitorShouldUpdateTheFetchOffsetInConsumer() {
+    // Define dummy variables for testing.
+    final Integer testPartitionId = 0;
+    final String offset = "0";
+    final TopicPartition testTopicPartition = new TopicPartition(TEST_STREAM, testPartitionId);
+    final Partition testPartition = new Partition(testPartitionId);
+    final SystemStreamPartition testSystemStreamPartition = new SystemStreamPartition(TEST_SYSTEM, TEST_STREAM, testPartition);
+
+    final KafkaConsumer consumer = Mockito.mock(KafkaConsumer.class);
+    KafkaStartpointVisitor kafkaStartpointVisitor = new KafkaStartpointVisitor(consumer);
+
+    final StartpointSpecific testStartpointSpecific = new StartpointSpecific(offset);
+
+    // Mock the consumer interactions.
+    Mockito.doNothing().when(consumer).seek(testTopicPartition, Long.valueOf(offset));
+
+    // Invoke the consumer with startpoint.
+    kafkaStartpointVisitor.visit(testSystemStreamPartition, testStartpointSpecific);
```

Comment:
If there is a public API for this functionality, it might be better to call that directly in the unit test.

Reply:
1. In general, we should try to test public APIs as much as possible instead of testing pieces of implementation. The production usage of our code is through public APIs, so it is best to test those directly. Sometimes, it is impractical to test public APIs, but in this case, it doesn't really seem any harder to test them directly. In addition, if someone adds logic in the future outside of the visitor, then the current tests won't catch that code.
2. Let's see what that other PR looks like, and then maybe we can update this PR based on that. This PR could just be committed after your other PR gets committed.

Reply:
1. Of-course, the plan is to test the public `SystemConsumer#register` API as well. But while testing it, I would just verify if the visitor mocks were invoked with right parameters. 
2. Yes, will send out the PR for it shortly.

Reply:
It doesn't seem necessary to have unit tests for both the public `register` API and the internal visitor methods, since those tests are going to cover pretty much the same thing.